### ISSUE TITLE
Can't delete a page of a type that was removed with migrations

### ISSUE THREAD

**Issue by alexgleason** [wagtail/wagtail#1951]
1. Create `TestPage(Page)` in models.py
2. makemigrations, migrate
3. Log into admin and create a TestPage page
4. Delete `TestPage(Page)` in models.py
5. makemigrations, migrate
6. Log into admin and delete test page from step 3

Result: Database error. TestPage isn't found in the database.

Expected result: The page is deleted gracefully.

**Comment by kaedroho** [comment]
I've seen this a couple of times too. I think there's a good chance that #2024 could solve it.

**Comment by gasman** [comment]
#2024 would only fix this if the ContentType record for the deleted page model gets deleted immediately, and I'm not sure we can rely on that (IIRC the "remove these stale content types?" prompt happens on `./manage.py migrate` some unspecified time later, and not at all if you run it with `--noinput`).

We have a check in Page.specific to catch this kind of scenario https://github.com/torchbox/wagtail/blob/aaed92f301a6862f30e817e23a50a21e8ae12e7b/wagtail/wagtailcore/models.py#L538-L544 - it sounds like this particular view is evading that check, and we probably need to repeat that check in some other places.

**Comment by atbaker** [comment]
We ran into a similar problem today which may be a symptom of the same issue:

We had a stale content type from breaking a particularly bloated snippet out into its own app. You're correct @gasman, that `migrate --noinput` will never remove stale content types for you.

For some reason, the groups edit page would 500 error while a stale content type for that snippet still existed. Here's the stacktrace:

```
Traceback (most recent call last):
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/django/views/decorators/cache.py", line 43, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/wagtail/wagtailadmin/decorators.py", line 22, in decorated_view
    return view_func(request, *args, **kwargs)
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/wagtail/wagtailadmin/utils.py", line 87, in wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/wagtail/wagtailusers/views/groups.py", line 133, in edit
    form = GroupForm(instance=group)
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/wagtail/wagtailusers/forms.py", line 197, in __init__
    self.registered_permissions = self.registered_permissions | fn()
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/wagtail/wagtailsnippets/wagtail_hooks.py", line 52, in register_permissions
    content_types = ContentType.objects.get_for_models(*get_snippet_models()).values()
  File "/Users/abaker/.virtualenvs/my-project/lib/python3.4/site-packages/django/contrib/contenttypes/models.py", line 114, in get_for_models
    if model._meta in needed_opts:
AttributeError: 'NoneType' object has no attribute '_meta'
```

We solved the problem by writing a database migration to remove the stale content type and applying it in all environments.

**Comment by danni** [comment]
Yep, I can confirm that you need to garbage collect stale content types, else the admin fails in obscure ways. Rerunning the migrate with input and removing the state content types downgraded the records to `Page` so they're deletable from the admin.

Some None detection when asking for content types would help here.

```
[05:29:35] PID:17(MainThread) django.template: [WARNING] Exception raised while rendering {% include %} for template 'wagtailadmin/pages/index.html'. Empty string rendered instead.
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/django/template/loader_tags.py", line 209, in render
    return template.render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/base.py", line 208, in render
    return self._render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/test/utils.py", line 92, in instrumented_test_render
    return self.nodelist.render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/loader_tags.py", line 173, in render
    return compiled_parent._render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/test/utils.py", line 92, in instrumented_test_render
    return self.nodelist.render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/usr/local/lib/python3.4/dist-packages/django/template/defaulttags.py", line 325, in render
    if match:
  File "/usr/local/lib/python3.4/dist-packages/django/core/paginator.py", line 117, in __len__
    return len(self.object_list)
  File "/usr/local/lib/python3.4/dist-packages/django/db/models/query.py", line 240, in __len__
    self._fetch_all()
  File "/usr/local/lib/python3.4/dist-packages/django/db/models/query.py", line 1074, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/usr/local/lib/python3.4/dist-packages/wagtail/wagtailcore/query.py", line 273, in specific_iterator
    pages = model.objects.filter(pk__in=pks)
AttributeError: 'NoneType' object has no attribute 'objects'
```

**Comment by gasman** [comment]
Fixed in #3569
